eppo_core/src/precomputed.rs

```diff
@@ -0,0 +1,180 @@
+    pub(crate) environment: Option<Environment>,
+    pub(crate) flags: HashMap<String, PrecomputedAssignment>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct PrecomputedAssignment {
+    pub(crate) variation_type: VariationType,
+    pub(crate) variation_value: ValueWire,
+
+    pub(crate) do_log: bool,
+    // If `do_log` is false, the client doesn’t need the field below.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub(crate) allocation_key: Option<Str>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub(crate) variation_key: Option<Str>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub(crate) extra_logging: Option<HashMap<String, String>>,
+}
+
+impl From<Assignment> for PrecomputedAssignment {
+    fn from(assignment: Assignment) -> PrecomputedAssignment {
+        match assignment.event {
+            Some(event) => PrecomputedAssignment {
+                variation_type: assignment.value.variation_type(),
+                variation_value: assignment.value.variation_value(),
+                do_log: true,
+                allocation_key: Some(event.base.allocation.clone()),
+                variation_key: Some(event.base.variation.clone()),
+                extra_logging: Some(event.base.extra_logging.clone()),
+            },
+            None => PrecomputedAssignment {
+                variation_type: assignment.value.variation_type(),
+                variation_value: assignment.value.variation_value(),
+                do_log: false,
+                allocation_key: None,
+                variation_key: None,
+                extra_logging: None,
+            },
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ObfuscatedPrecomputedConfiguration {
+    obfuscated: serde_bool::True,
+    /// `format` is always `AssignmentFormat::Precomputed`.
+    format: ConfigurationFormat,
+    /// Salt used for hashing md5-encoded strings.
+    #[serde_as(as = "serde_with::base64::Base64")]
+    salt: [u8; 16],
+    created_at: Timestamp,
+    // Environment might be missing if configuration was absent during evaluation.
+    environment: Option<Environment>,
+    flags: HashMap<Md5HashedStr, ObfuscatedPrecomputedAssignment>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(crate) struct ObfuscatedPrecomputedAssignment {
+    variation_type: VariationType,
+    variation_value: Base64Str,
+    do_log: bool,
+    // If `do_log` is false, the client doesn’t need the fields below.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    allocation_key: Option<Base64Str>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    variation_key: Option<Base64Str>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    extra_logging: Option<HashMap<Base64Str, Base64Str>>,
+}
```

Comment:
@typotter @sameerank This is the format for obfuscated configuration that edge function is going to produce. Hope it's easy to see where md5 (`Md5HashedStr`) or base64 (`Base64Str`) is used.

All md5 hashes are computed by prepending `ObfuscatedPrecomputedConfiguration::salt` to the string. It's a 16 random bytes encoded as base64 in the config. We need to base64-decode it before hashing. Note that because it's random bytes and not a string, [`btoa` may fail](https://developer.mozilla.org/en-US/docs/Glossary/Base64#javascript_support) and we need to use array version of decoding in js

Reply:
Here's an example for flags in sdk-test-data / flags-v1.json:
<details>

```json
{
  "obfuscated": true,
  "format": "PRECOMPUTED",
  "salt": "/Ls8/hc2x1wDOKQ33syRnw==",
  "created_at": "2024-12-06T17:36:11.116639Z",
  "environment": {
    "name": "Test"
  },
  "flags": {
    "ecf30449f115fa08881ed49d4124749a": {
      "variation_type": "NUMERIC",
      "variation_value": "My4xNDE1OTI2",
      "do_log": true,
      "allocation_key": "cm9sbG91dA==",
      "variation_key": "cGk=",
      "extra_logging": {}
    },
    "3bab431fe39dc3cdd6831b3c6ec1e6a3": {
      "variation_type": "STRING",
      "variation_value": "b2xk",
      "do_log": true,
      "allocation_key": "bnVsbC1vcGVyYXRvcg==",
      "variation_key": "b2xk",
      "extra_logging": {}
    },
    "550bb44d14f53e13eb6a1f1082c20c0f": {
      "variation_type": "JSON",
      "variation_value": "eyAiaW50ZWdlciI6IDIsICJzdHJpbmciOiAidHdvIiwgImZsb2F0IjogMi4wIH0=",
      "do_log": true,
      "allocation_key": "NTAvNTAgc3BsaXQ=",
      "variation_key": "dHdv",
      "extra_logging": {}
    },
    "ec055680f8e0b41f47b70cf15a8c175f": {
      "variation_type": "INTEGER",
      "variation_value": "MQ==",
      "do_log": true,
      "allocation_key": "NTAvNTAgc3BsaXQ=",
      "variation_key": "b25l",
      "extra_logging": {}
    },
    "24bcd66520102821e592040a24960ec0": {
      "variation_type": "BOOLEAN",
      "variation_value": "ZmFsc2U=",
      "do_log": true,
      "allocation_key": "b2ZmLWZvci1hbGw=",
      "variation_key": "b2Zm",
      "extra_logging": {}
    },
    "b0a851a2072ee5278ced2680c22e4627": {
      "variation_type": "STRING",
      "variation_value": "Y3VycmVudA==",
      "do_log": true,
      "allocation_key": "Y3VycmVudC12ZXJzaW9ucw==",
      "variation_key": "Y3VycmVudA==",
      "extra_logging": {}
    }
  }
}
```

</details>

Reply:
A few thoughts from looking at the example (thanks for putting that together!):

1. Looks like I need to add `variationType` to the decoded items in https://github.com/Eppo-exp/js-sdk-common/blob/main/src/decoding.ts#L81-L88
2. We've been using camelCase for the UFC and so far also for the precomputed object, e.g. https://github.com/Eppo-exp/js-sdk-common/blob/d994780e0916520700bea5947085f38f5d0c677a/src/client/eppo-precomputed-client.spec.ts#L18-L66. Not generally opposed to snake_case, but it would be inconsistent in this situation
3. Could be worth having some key-values in `extraLogging` to illustrate `decodeObject` in action
4. I also need to update https://github.com/Eppo-exp/js-sdk-common/blob/d994780e0916520700bea5947085f38f5d0c677a/src/client/eppo-precomputed-client.ts#L278-L280 to use the `salt` when hashing the flag key

Reply:
Good catch about `camelCase` — it should be `camelCase`, I just forgot the annotation.

Reply:
For `extraLogging`, both keys and values are base64-encoded:
```python
{
  ...,
  "extraLogging": {
    "aGVsbG8K": "d29ybGQK",
  }
}
```

Reply:
I've written out how this will look within the "configuration wire" format here (as I'm understanding it): https://github.com/Eppo-exp/sdk-test-data/pull/89/files

Reply:
I think the one question that came to mind in the process is if this is the right level to nest the `obfuscated` boolean. Will `subjectKey` and `subjectAttributes` (which are in the parent object) ever be in an obfuscated format and can we use the child object's attribute to determine if they are obfuscated?